src/Storages/PostgreSQL/PostgreSQLReplicaConsumer.cpp

```diff
@@ -0,0 +1,682 @@
+                        "Storage for table {} does not exist, but is included in replication stream", relation_name);
+            }
+
+            [[maybe_unused]] auto buffer_iter = buffers.find(relation_name);
+            assert(buffer_iter != buffers.end());
+
+            tables_to_sync.insert(relation_name);
+
+            break;
+        }
+        case 'O': // Origin
+            break;
+        case 'Y': // Type
+            break;
+        case 'T': // Truncate
+            break;
+        default:
+            throw Exception(ErrorCodes::LOGICAL_ERROR,
+                    "Unexpected byte1 value {} while parsing replication message", type);
+    }
+}
+
+
+void PostgreSQLReplicaConsumer::syncTables(std::shared_ptr<pqxx::nontransaction> tx)
+{
+    for (const auto & table_name : tables_to_sync)
+    {
+        try
+        {
+            auto & buffer = buffers.find(table_name)->second;
+            Block result_rows = buffer.description.sample_block.cloneWithColumns(std::move(buffer.columns));
+
+            if (result_rows.rows())
+            {
+                metadata.commitMetadata(final_lsn, [&]()
+                {
+                    auto storage = storages[table_name];
+
+                    auto insert = std::make_shared<ASTInsertQuery>();
+                    insert->table_id = storage->getStorageID();
+                    insert->columns = buffer.columnsAST;
+
+                    auto insert_context(*context);
+                    insert_context.makeQueryContext();
+                    insert_context.addQueryFactoriesInfo(Context::QueryLogFactories::Storage, "ReplacingMergeTree");
+
+                    InterpreterInsertQuery interpreter(insert, insert_context, true);
+                    auto block_io = interpreter.execute();
+                    OneBlockInputStream input(result_rows);
+
+                    assertBlocksHaveEqualStructure(input.getHeader(), block_io.out->getHeader(), "postgresql replica table sync");
+                    copyData(input, *block_io.out);
+
+                    auto actual_lsn = advanceLSN(tx);
+                    buffer.columns = buffer.description.sample_block.cloneEmptyColumns();
+
+                    return actual_lsn;
+                });
+            }
+        }
+        catch (...)
+        {
+            tryLogCurrentException(__PRETTY_FUNCTION__);
+        }
+    }
+
+    LOG_DEBUG(log, "Table sync end for {} tables", tables_to_sync.size());
+    tables_to_sync.clear();
+    tx->commit();
+}
+
+
+String PostgreSQLReplicaConsumer::advanceLSN(std::shared_ptr<pqxx::nontransaction> tx)
+{
+    std::string query_str = fmt::format("SELECT end_lsn FROM pg_replication_slot_advance('{}', '{}')", replication_slot_name, final_lsn);
+    pqxx::result result{tx->exec(query_str)};
+
+    if (!result.empty())
+        return result[0][0].as<std::string>();
+
+    return final_lsn;
+}
+
+
+bool PostgreSQLReplicaConsumer::isSyncAllowed(Int32 relation_id)
```

Comment:
Please add a comment with explanation why sync is allowed or not in different cases.

Reply:
It is explained in this file: https://github.com/ClickHouse/ClickHouse/blob/caffccd73e068ae2aea43a447426aee217694d39/src/Storages/PostgreSQL/MaterializePostgreSQLConsumer.h#L138

Reply:
> Please add a comment with explanation why sync is allowed or not in different cases.

Added.